sklearn/feature_selection/tests/test_mutual_info.py

```diff
@@ -0,0 +1,173 @@
+
+        np.random.seed(0)
+        Z = np.random.multivariate_normal(mean, cov, size=1000)
+
+        x, y = Z[:, 0], Z[:, 1]
+        I_computed = _compute_mi(x, y, False, False)
+
+        # Theory and computed values won't be very close, assert that relative
+        # error is less than 10%.
+        assert_true(np.abs(I_computed - I_theory) < 0.1 * I_theory)
+
+    def test_cd(self):
+        # To test define a joint distribution as follows:
+        # p(x, y) = p(x) p(y | x)
+        # X ~ Bernoulli(0.5)
+        # (Y | x = 0) ~ Uniform(-1, 1)
+        # (Y | x = 1) ~ Uniform(0, 2)
+
+        # Use the following formula for mutual information:
+        # I(X; Y) = H(Y) - H(Y | X)
+        # Two entropies can be computed by hand:
+        # H(Y) = 3/2 * ln(2)
+        # H(Y | X) = ln(2)
+        # So finally we have:
+        # I(X; Y) = 1/2 * ln(2)
+
+        # Now we need to implement sampling from out distribution, which is
+        # done easily using conditional distribution logic.
+
+        n_samples = 1000
+        np.random.seed(0)
+        Z = np.empty((n_samples, 2))
+        Z[:, 0] = np.random.uniform(size=n_samples) > 0.5
+        mask = Z[:, 0] == 0
+        Z[mask, 1] = np.random.uniform(-1, 1, size=np.sum(mask))
+        mask = ~mask
+        Z[mask, 1] = np.random.uniform(0, 2, size=np.sum(mask))
+
+        x = Z[:, 0].astype(int)
+        y = Z[:, 1]
+
+        I_theory = 0.5 * np.log(2)
+        I_computed = _compute_mi(x, y, True, False)
+
+        # Again assert that relative error is less than 10%.
+        assert_true(np.abs(I_computed - I_theory) < 0.1 * I_theory)
+
+
+class TestMinRedundancyMaxRelevance(object):
+    # We'll have to rely on "common sense" to test our procedure. That is
+    # test on data where we can judge which features are "most relevant and
+    # least redundant".
+    def test_categorical(self):
+        X = np.array([
+            [0, 0, 0, 0],
+            [1, 1, 0, 1],
+            [2, 2, 1, 0],
+            [2, 2, 1, 0],
+            [2, 2, 1, 0]
+        ], dtype=float)
+        y = np.array([0, 1, 2, 2, 1])
+
+        # Here x[0] is the most informative feature, x[1] is its duplicate
+        # (thus redundant) and x[3] is weekly informative. So the algorithm
+        # should pick features 0 and 2.
+
+        m = MinRedundancyMaxRelevance(
+            categorical_features=True, categorical_target=True,
+            n_features_to_select=2)
+        m.fit(X, y)
+        assert_array_equal(m.support_, np.array([True, False, True, False]))
+
+    def test_continuous(self):
+        # The idea is to generate sample from multivariate normal distribution,
+        # with interesting correlation relations. Here the target variable
+        # 1 is strongly correlated to the variables 2 and 3. But the
+        # variables 2 and 3 are strongly correlated to each other, and thus
+        # the algorithm should prefer the 3rd variable to the 2nd variable.
```

Comment:
But it picks the 2nd variable. This is logical since Cov(X1, X2) > Cov(X1, X3) . I think this is just a mistake while documenting.